### PR TITLE
Add new module to convert cp stats into prometheus format suited for CMS monitoring

### LINKED ISSUE #9939: Provide proper metrics for web based services
**Impact of the new feature**
Based on recent incident [INC2542764](https://cern.service-now.com/service-portal?id=ticket&table=incident&n=INC2542764) with low service performance of Python based services on k8s it would be extremely helpful to provide proper set of metrics from service stack. Since all cmsweb Python based services (DBS, ReqMgr2, CRAB etc.) are based on WMCore codebase it is natural to implement service metrics in WMCore rather in upstream services. This will allow proper monitoring of web services.

**Describe the solution you'd like**
Extend [ProcessStats.py](https://github.com/dmwm/WMCore/blob/master/src/python/Utils/ProcessStats.py) metrics to include the following
- number of requests
- status code stats
- response time of request
- number of connections
- number of open file descriptors
- if DB is in use DB related stats

**Is your feature request related to a problem? Please describe.**
Please see this incident discusssions [INC2542764](https://cern.service-now.com/service-portal?id=ticket&table=incident&n=INC2542764)

**Describe alternatives you've considered**
None, currently we run our services without proper knowledge how they behave, what is their throughput and how the perform under production load

**Additional context**
Once metrics are added to [ProcessStats.py](https://github.com/dmwm/WMCore/blob/master/src/python/Utils/ProcessStats.py) we need to modify WMCore WebTools and REST interface to add `/metrics` or `/status` end-points and enable this by default that it would not be required in upstream code to change anything in order to have them. Once they are available we can setup proper exporter (code should be developed in CMS Monitoring group) and integrate it into Prometheus metrics for each individual service.

### PR DESCRIPTION
**Pull request by vkuznet**
Fixes #9939 

#### Status
ready

#### Description
I provide a new module which can be integrated into WMCore stack. It provides useful set of functions to flatten CherryPy stats, make static schema and provide them in format suitable for Prometheus server.

I suggest to add an additional `metrics` endpoint to complement `stats` one shown in here
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/REST/Server.py#L700
with the following
```
@expose
def metrics(self):
       name = self.__class__.__name__
       return prom_metrics(name)
```

This will provide ability to scrape all cherrypy metrics from Prometheus for any DMWM/REST server. Please refer to docstring of provided module for format description/example.

#### Is it backward compatible (if not, which system it affects?)
YES

#### External dependencies / deployment changes
this module uses cherrypy lib to get cherrypy metrics

### CODE REVIEW

**Review by amaltaro — CHANGES_REQUESTED** [review]
Not really a review of the code you proposed, but a couple of things caught my attention:
* please use lower camelCase for the variable and function names (prom_metrics --> promMetrics)
* given that this is going to be a common Utils module, please provide docstrings as completed as you can (there is no model defined yet, but it looks like we are adopting reStructured Text)
* can you also please create unit test(s) for this module

Can you please take care of the comments above? Thanks

**Comment by vkuznet** [comment]
Alan, I added requested changes, please review at your convenience.

File: src/python/Utils/CPMetrics.py
Comment (todor-ivanov) [inline src/python/Utils/CPMetrics.py:258]: Same comment as above regarding the variable names for `wkey` and `tval`. Since here it already becomes clear there is a third level in the CherryPy data structure we are considering, I would, for sure, make those variable names more descriptive.

File: src/python/Utils/CPMetrics.py
Comment (todor-ivanov) [inline src/python/Utils/CPMetrics.py:276]: A typo here. Should be:
`     Example of metrics server which can serve cherrypy metrics`

**Review by todor-ivanov — APPROVED** [review]
Valya, I made few minor comments in the code. But, if you think those are too picky, just ignore them, we can definitely live with the code as it is. 
All the rest looks good to me!

File: src/python/Utils/CPMetrics.py
Comment (vkuznet) [reply src/python/Utils/CPMetrics.py:245]: Todor, it is each to check, it work fine with python3
```
Python 3.7.9 (default, Sep  6 2020, 13:20:25)
[Clang 11.0.3 (clang-1103.0.32.62)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> s="bla"
>>> isinstance(s, str)
True
```

File: src/python/Utils/CPMetrics.py
Comment (vkuznet) [reply src/python/Utils/CPMetrics.py:248]: Since we don't own CP stats data structure we cannot guarantee that it will be sustainable. It is their API at the end. But so far the parser I provided works with current CP stats structure.

Comment (vkuznet) [reply src/python/Utils/CPMetrics.py:248]: regarding keys/vals, I changed them as you requested

File: src/python/Utils/CPMetrics.py
Comment (vkuznet) [reply src/python/Utils/CPMetrics.py:258]: ok, replaced to workerKey and threadValue.

File: src/python/Utils/CPMetrics.py
Comment (vkuznet) [reply src/python/Utils/CPMetrics.py:276]: done

**Comment by vkuznet** [comment]
@todor-ivanov  please review again as I resolved all requested suggestions.

File: src/python/Utils/CPMetrics.py
Comment (mapellidario) [reply src/python/Utils/CPMetrics.py:245]: In this case, we may be luckier that expected and no modification may be required.

#### Fact 1

`json.loads()` is very permissive in the type of its input.

1. [python2](https://docs.python.org/2/library/json.html#json.loads): `str` or `unicode`
2. [python3](https://docs.python.org/3/library/json.html#json.loads): `str`, `bytes` or `bytearray`

If  `cpdata` is of type `str`, then it is valid input for `json.loads()` in both python2 and python3.

`json.dumps()` return a `str` in both [py2](https://docs.python.org/2/library/json.html#json.dumps) (meaning a sequence of bytes here) and [py3](https://docs.python.org/3/library/json.html#json.dumps) (meaning a sequence of unicode codepoints). 

#### Fact 2

I am not an expert of `cherrypy`, but looking at [pypi](https://pypi.org/project/CherryPy/) the developers state that 

1. version [17.4.2](https://pypi.org/project/CherryPy/17.4.2/) of cherrypy is the last compatible with both py2 (py2.7) and py3 (py3.0->py3.3)
2. version [18.6.0](https://pypi.org/project/CherryPy/18.6.0/) of cherrypy (latest at time fo this comment) is compatible with py3 only (>= py3.5)

In version [17.4.2](https://github.com/cherrypy/cherrypy/blob/c3c2cec9a74f2c9fe71849c9aff5ab0e7f5743ec/cherrypy/lib/cpstats.py#L679), `cherrypy.lib.cpstats.StatsPage.data()` returns `json.dumps()`. If we run this in python2, then `cpdata` is of type `str` (meaning a sequence of bytes) and it is valid input for `json.loads()`.

In version [18.6.0](https://github.com/cherrypy/cherrypy/blob/b2938bc92570a1a04aa4336412e1824b9fe455eb/cherrypy/lib/cpstats.py#L677), `cherrypy.lib.cpstats.StatsPage.data()` returns `json.dumps().encode('utf-8')`. If we run this in python3, then `cpdata` is of type `str` (meaning a sequence of unicode codepoints) and is valid input for `json.loads()`

Comment (todor-ivanov) [reply src/python/Utils/CPMetrics.py:245]: Thanks @vkuznet and @mapellidario !

**Review by todor-ivanov — APPROVED** [review]
Hi Valya, I think it is looking good now. Just to clear my own understanding. From your comment in the PR I read that provided the code like that it is not referred anywhere outside. So we should make an explicit call to it from here: https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/REST/Server.py#L700

`
@expose
def metrics(self):
       name = self.__class__.__name__
       return prom_metrics(name)
 `

If I am correct it would be nice to have it called in this same PR so that  we avoid making too many PRs for the  same functionality. @amaltaro, Please take a look and tell me if you like to have the code merged with or without being referred in this PR.

**Comment by vkuznet** [comment]
Todor,
it is fine from my side to add changes to REST Server in this PR. But I think either you or Alan should decide on it. Here is a proposed changes (in diff format):
```
diff --git a/src/python/WMCore/REST/Server.py b/src/python/WMCore/REST/Server.py
index b3a3b8a73..450040e49 100644
--- a/src/python/WMCore/REST/Server.py
+++ b/src/python/WMCore/REST/Server.py
@@ -15,6 +15,7 @@ from cherrypy.lib import cpstats
 from WMCore.REST.Error import *
 from WMCore.REST.Format import *
 from WMCore.REST.Validation import validate_no_more_input
+from Utils.CPMetrics import promMetrics

 try:
     from cherrypy.lib import httputil
@@ -360,6 +361,11 @@ class RESTFrontPage:
         "Return CherryPy stats dict about underlying service activities"
         return cpstats.StatsPage().data()

+    @expose
+    def metrics(self):
+        "Return CherryPy stats metrics in Prometheus format"
+        name = self.__class__.__name__
+        return promMetrics(name)


 ######################################################################
```

**Comment by amaltaro** [comment]
Valentin, just to let you know that this PR is on my todo list for the coming days. I still think there is either something missing or a misunderstanding with this proposal, so I'll have to test a few things myself first.

**Comment by vkuznet** [comment]
@amaltaro , I incorporated your changes from #9961 over here. I removed cherrypy deps and TestMetricsServer from CPMetrics.py code. But it would be nice to keep TestMetricsServer some place that users can easily see how to construct required end-point and test it easily. I created this file but did not include it in this PR. If you think it is relevant let me know where to put it, otherwise we may just skip it.
```
#!/usr/bin/env python
"""
File       : cp_metrics.py
Author     : Valentin Kuznetsov <vkuznet AT gmail dot com>
Description: Example of Python REST Server to provide cherrypy stats as prometheus metrics
"""



import cherrypy
from cherrypy.lib import cpstats

class MetricsServer(object):
    """
    Example of metrics server which can serve cherrypy metrics
    in Prometheus format
    """
    def metrics(self):
        "metrics end-point for prometheus to scrape"
        data = cpstats.StatsPage().data()
        pdata = promMetrics(data, 'test')
        return pdata
    metrics.exposed = True

    def index(self):
        "default index end-point"
        return "Hello World"
    index.exposed = True


def TestMetricsServer():
    "Test MetricsServer function"
    cherrypy.root = MetricsServer()
    cherrypy.config.update({'server.socket_port': 8080,
                            'server.thread_pool': 20,
                            'environment': 'production',
                            'log.screen': True,
                            'log.error_file': "crab.log"})
    conf = {'/': {'tools.staticdir.root': os.getcwd()}}
    cherrypy.quickstart(MetricsServer(), '/', config=conf)

if __name__ == '__main__':
    TestMetricsServer()

```

**Comment by vkuznet** [comment]
@amaltaro , I found where stats and metrics can be plugged into WebTools server. Looking at DBS code it inherits WebTools RESTModel, which by itself inherits from WebAPI.py. The whole chain of dependencies is the following:
- https://github.com/dmwm/DBS/blob/master/Server/Python/src/dbs/web/DBSReaderModel.py#L69
- https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WebTools/RESTModel.py#L24
- https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WebTools/WebAPI.py

So, I added stats and metrics to WebAPI and exposed them. Doing this way all DBS servers (inherited from WebTools) will get these end-points.

Please review.

**Comment by vkuznet** [comment]
I had a look at pylint report and all of the failures related to WebAPI.py codebase which I didn't touched. I don't think it is my responsibility to go through them and include in this PR. Please let me know how you'll treat it.

**Comment by vkuznet** [comment]
I had a look at Alan's VM and even though I can access provided end-points via https calls they are not accessible from localhost and plain http:
```
# https requests works fine
https://alancc7-cloud1.cern.ch/reqmgr2/data/metrics

# http request is not allowed giving 403 forbidden page
http://localhost:8246/reqmgr2/data/metrics
```

Such strict constrains make entire work a waste since prometheus does not care about cmsweb auth x509 schema and all metrics should be accessible through normal localhost request. Therefore, unless I miss something or unless you tell me where to change REST server code to allow http requests from localhost I don't see a point how it can be used because REST server requires cmsweb x509 authentication for all exposed methods.

**Review by todor-ivanov — APPROVED** [review]
This version seems to properly incorporate both the initial idea and Alan's requirements. Thanks @vkuznet and @amaltaro. 

Valya, just a minor confirmation here. Those lines, you did test them for real with a Prometheus instance, right? Or it was just the code from the WMcore side that you tested?

**Comment by vkuznet** [comment]
Todor, I tested all changes with TestMetricsServer, the code I provided in original PR content or I put it in this thread, see https://github.com/dmwm/WMCore/pull/9940#issuecomment-702801325

But I also commented that I can't test it properly with WMCore services since there is no localhost access to them and the only access it through external frontend which requires X509 authentication, e.g. `https://alancc7-cloud1.cern.ch/reqmgr2/data/metrics`. The URL works with curl and my certificates, but since Prometheus does not use x509 authentication I can't use/test it. And, call to `http://localhost:8246/reqmgr2/data/metrics` is forbidden on the host where it runs.

**Comment by todor-ivanov** [comment]
Thanks @vkuznet !

**Comment by amaltaro** [comment]
@vkuznet Valentin, trying to move forward with this PR. Are you able to talk to Imran and try to figure out how the ServerMonitor manages to make localhost calls to every single service under CMSWEB?

**Comment by amaltaro** [comment]
@vkuznet Valentin, I think these localhost queries are made with some specific http headers. Have you figured that out or have you found a way to make it working with prometheus?

**Comment by vkuznet** [comment]
Alan, the [ServerMonitor](https://github.com/dmwm/deployment/blob/master/admin/ServerMonitor#L27) and any calls to localhost of WMCore services requires proper hmac file and CMS headers. This is constrain of WMCore REST code, see [Auth.py](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/REST/Auth.py).  WIthout CMS headers all calls to WMCore REST based services is prohibited. As such Prometheus can't talk to local host of the service using /metrics end-point since it does not have any knowledge of our authz schema. And, I doubt that we should modify prometheus to handle CMS headers. I rather want to see exception in WMCore REST  implementation which will allow calls from localhost.

**Comment by vkuznet** [comment]
test this please

**Comment by vkuznet** [comment]
@amaltaro I revisited this PR and adjusted it to pass unit tests, etc. It still shows plenty of pylint errors in [src/python/WMCore/REST/Server.py]() but I doubt I should fix them The REST server codebase is outside the scope of this PR and since it has (by itself) so many different issues with pylint check I don't want to fix them here.

I would appreciate if you finally, after two years :), will review this PR as we plan to include it as part of https://github.com/dmwm/WMCore/issues/9939 which we claim on Q2/22, and let me know if any other action items we need on this.

**Comment by amaltaro** [comment]
@vkuznet thanks for ping'in on this.
I left a couple of comments that we need to update in the code, and this will allow us to see how unit tests behave after such a long time. Please make sure to keep src/ changes separated from test.

Once you make this new commit, can you please rebase this branch to see the actual changes? Thanks

**Review by amaltaro — COMMENTED** [review]
My comment above was made at the same time as the review request, so please see my feedback above.

File: test/python/Utils_t/CPMetrics_t.py
Comment (vkuznet) [reply test/python/Utils_t/CPMetrics_t.py]: fixed

**Comment by vkuznet** [comment]
@amaltaro , I rebased the code. Please proceed with review.

File: src/python/Utils/CPMetrics.py
Comment (amaltaro) [inline src/python/Utils/CPMetrics.py:187]: Sorry, I missed this. We should target python3 only, so please remove this import and make sure that division is behaving properly without it.

File: test/python/Utils_t/CPMetrics_t.py
Comment (amaltaro) [inline test/python/Utils_t/CPMetrics_t.py:6]: Same comment here. We no longer require such compatibility changes.

File: src/python/Utils/CPMetrics.py
Comment (amaltaro) [inline src/python/Utils/CPMetrics.py:193]: please stick to lower camel case for variable names.

**Review by amaltaro — CHANGES_REQUESTED** [review]
Thanks for these changes, Valentin. I left a few comments along the code (to remove py2/py3 compatibility import).

Before this can go through, please:
* check the pylint and pycodestyle for the new modules. Those should be as clean as possible.
* unit tests are likely unstable. So we can just trigger the tests again to see if they succeed
* commits need to be correctly squashed

File: src/python/Utils/CPMetrics.py
Comment (vkuznet) [reply src/python/Utils/CPMetrics.py]: done

File: src/python/Utils/CPMetrics.py
Comment (vkuznet) [reply src/python/Utils/CPMetrics.py]: done

File: test/python/Utils_t/CPMetrics_t.py
Comment (vkuznet) [reply test/python/Utils_t/CPMetrics_t.py]: done

**Comment by vkuznet** [comment]
Alan, pylint score is 10 for new code and its unit test. All others are old code which I don't want to touch in this PR. The unit tests are unstable. Please review once you'll get a chance.

**Review by amaltaro — APPROVED** [review]
This looks good to me.
Before we get it merged, can you please deploy these changes to your test9 cluster and run a basic curl test of the new REST API?

**Comment by vkuznet** [comment]
@amaltaro , after testing on test9 with curl I found that additional changes to support `bytes` I/O is required (new python CP server should return bytes for its end-points). So I added separte commit https://github.com/dmwm/WMCore/pull/9940/commits/3813d7eb6def6cca6a859f81946329ef28354e01 Please review this commit and then I can squash it.

File: src/python/Utils/CPMetrics.py
Comment (amaltaro) [inline src/python/Utils/CPMetrics.py:239]: To load this json you need to have unicode data. I would suggest to use this utilitarian method instead:
https://github.com/dmwm/WMCore/blob/master/src/python/Utils/Utilities.py#L203

Can you please replace this if/else by the following line:
```
cpdata = json.loads(decodeBytesToUnicode(cpdata))
```
?

**Review by amaltaro — CHANGES_REQUESTED** [review]
Valentin, please  see my partial review on your latest changes.

File: src/python/Utils/CPMetrics.py
Comment (vkuznet) [reply src/python/Utils/CPMetrics.py]: done

File: src/python/WMCore/REST/Server.py
Comment (vkuznet) [reply src/python/WMCore/REST/Server.py]: done

**Comment by vkuznet** [comment]
Alan, I made necessary changes, please let me know.

**Comment by vkuznet** [comment]
Now commits are squashed too.

**Comment by amaltaro** [comment]
The Jenkins CI report is your friend, please take advantage of it ;)
```
Unit tests added
    Utils_t.CPMetrics_t.CPMetricsTests:testFlattenStats was added with status error. Must be fixed
    Utils_t.CPMetrics_t.CPMetricsTests:testPromMetrics was added with status error. Must be fixed
```

Once you are done with code changes, please run another test in your central services setup testing reqmgr2 endpoints like: stats, metrics and info. Please let me know how that goes.

**Comment by vkuznet** [comment]
Alan, test is fixed (adding encodeUnicodetoBytes still requires a check) and code is tested on test9, e.g.
```
scurl -s https://cmsweb-test9.cern.ch/ms-output/data/metrics | head
# HELP ms-output_cherrypy_http_server_accepts
# TYPE ms-output_cherrypy_http_server_accepts counter
ms-output_cherrypy_http_server_accepts 0
# HELP ms-output_cherrypy_http_server_accepts_sec
# TYPE ms-output_cherrypy_http_server_accepts_sec gauge
ms-output_cherrypy_http_server_accepts_sec 0.0
# HELP ms-output_cherrypy_http_server_bytes_read
# TYPE ms-output_cherrypy_http_server_bytes_read counter
ms-output_cherrypy_http_server_bytes_read -1
# HELP ms-output_cherrypy_http_server_bytes_written
```

**Review by amaltaro — APPROVED** [review]
Looks good to me, thanks!

**Comment by mapellidario** [comment]
From a first quick pick, this is beautiful! We had the idea of doing something similar and will likely steal from this in the future :)
Fyi: @novicecpp

**Comment by vkuznet** [comment]
FYI @amaltaro , @mapellidario , @novicecpp I made corresponding monitoring ticket: https://its.cern.ch/jira/browse/CMSMONIT-514 to request CherryPy dashboard. Please follow up monitoring ticket and request your dashboard for your python based service.

**Pull request merged**
